File: Python3/프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    message_log = []
    user_info = {}

    for entry in record:
        parts = entry.split()
        actions = parts[0]
        user_id = parts[1]

        if actions == "Enter":
            nickname = parts[2]
            user_info[user_id] = nickname
            message_log.append((user_id, "님이 들어왔습니다."))
        elif actions == "Leave":
            message_log.append((user_id, "님이 나갔습니다."))
        elif actions == "Change":
            nickname = parts[2]
            user_info[user_id] = nickname

    final_messages = []

    for user_id, message in message_log:
        final_messages.append(user_info[user_id] + message)

    return final_messages

File: Swift/프로그래머스/2/42888. 오픈채팅방/오픈채팅방.swift
import Foundation

func solution(_ record:[String]) -> [String] {
    var messageLog: [(String, String)] = []
    var userInfo: [String : String] = [:]
    
    for entry in record {
        let parts = entry.split(separator: " ").map{ String($0) }
        let action = parts[0]
        let userId = parts[1]
        
        if action == "Enter" {
            let nickname = parts[2]
            userInfo[userId] = nickname
            messageLog.append((userId, "님이 들어왔습니다."))
        } else if action == "Leave"{
            messageLog.append((userId, "님이 나갔습니다."))
        } else if action == "Change" {
            let nickname = parts[2]
            userInfo[userId] = nickname
        }
    }
    
    var finalMessages: [String] = []
    
    for log in messageLog {
        let userId = log.0
        let message = log.1
        if let nickname = userInfo[userId] {
            finalMessages.append("\(nickname)\(message)")
        }
    }
    
    return finalMessages
}
